db/types.go

```diff
@@ -0,0 +1,183 @@
+package db
+
+import "errors"
+
+var (
+	// ErrBatchClosed is returned when a closed or written batch is used.
+	ErrBatchClosed = errors.New("batch has been written or closed")
+
+	// ErrKeyEmpty is returned when attempting to use an empty or nil key.
+	ErrKeyEmpty = errors.New("key cannot be empty")
+
+	// ErrValueNil is returned when attempting to set a nil value.
+	ErrValueNil = errors.New("value cannot be nil")
+
+	// ErrVersionDoesNotExist is returned when a DB version does not exist.
+	ErrVersionDoesNotExist = errors.New("version does not exist")
+
+	// ErrOpenTransactions is returned when open transactions exist which must
+	// be discarded/committed before an operation can complete.
+	ErrOpenTransactions = errors.New("open transactions exist")
+
+	// ErrReadOnly is returned when a write operation is attempted on a read-only transaction.
+	ErrReadOnly = errors.New("cannot modify read-only transaction")
+)
+
+// DBConnection represents a connection to a versioned database.
+// Records are accessed via transaction objects, and must be safe for concurrent creation
+// and read and write access.
+// Past versions are only accessible read-only.
+type DBConnection interface {
+	// Opens a read-only transaction at the current working version.
+	Reader() DBReader
+
+	// Opens a read-only transaction at a specified version.
+	// Returns ErrVersionDoesNotExist for invalid versions.
+	ReaderAt(uint64) (DBReader, error)
+
+	// Opens a read-write transaction at the current version.
+	ReadWriter() DBReadWriter
+
+	// Opens a write-only transaction at the current version.
+	Writer() DBWriter
+
+	// Returns all saved versions
+	Versions() VersionSet
+
+	// Saves the current contents of the database and returns the next ID.
+	// If 0 is passed as target, the default version ID used will be `Versions().Last()+1`.
```

Comment:
Instead of this behavior, can we just have two methods: `SaveNextVersion() (uint64, error)` and `SaveVersion(uint64) error)`?